### PR TITLE
Pre & Post commands 

### LINKED ISSUE #453: Support Pre & Post build scripts
## Context
I was creating a fiber API and wanted to run the `swag fmt` and `swag init` commands before each build.

## Wanted Solution
```config
[build]
    pre_cmd = [ "swag fmt", "swag init" ] 
    cmd = "go build -o ./tmp/main.exe ."
    post_cmd = [ " some other scripts " ]
```

If the maintainers are ok with this you can assign this to me and I will happily work on it :smile:

### PR DESCRIPTION
**Pull request by kareemmahlees**
closes #453 
## New Features
- A new option `pre_cmd` is added to `.air.toml`
- A new option `post_cmd` is added to `.air.tom`

## New behavior
- commands that are in `pre_cmd` will be executed one by one before each build
- commands that are in `post_cmd` will be executed upon exiting (i.e. upon hitting ^C)

### CODE REVIEW

**Comment by kareemmahlees** [comment]
@xiantang could you review the changes made until now

**Comment by xiantang** [comment]
So, this time is it in support of the pre command?

File: runner/engine_test.go [hunk 1]
```diff
@@ -185,6 +185,28 @@ func TestRerunWhenFileChanged(t *testing.T) {
...
+		t.Fatalf("Should not be fail: %s.", err)
+	}
+}
+
+func TestRunPreCmd(t *testing.T) {
+	engine, err := NewEngine("", true)
+	if err != nil {
+		t.Fatalf("Should not be fail: %s.", err)
+	}
+	err = engine.runPreCmd()
```
Comment (xiantang) [inline runner/engine_test.go:204]: Unit tests should not be written like this. You need to check if the corresponding values have changed

**Comment by kareemmahlees** [comment]
Yes, will work on the post command shortly

**Comment by kareemmahlees** [comment]
> So, this time is it in support of the pre command?

Yes, will work on the post command shortly.

File: runner/engine_test.go (hunk 1 above)
Comment (kareemmahlees) [reply runner/engine_test.go:204]: I don't think I completely understand you!

Comment (xiantang) [inline runner/engine_test.go:204]: You should use a real command for integration testing, such as writing a file to the test folder and then checking if the file exists after execution is complete

Comment (xiantang) [reply runner/engine_test.go:204]: I don't think the current unit tests can test anything

File: air_example.toml [hunk 2]
```diff
@@ -6,6 +6,8 @@ root = "."
 tmp_dir = "tmp"
 
 [build]
+# Array of commands to run before each build
+pre_cmd = ["echo Hello","echo Air"]
```
Comment (xiantang) [inline air_example.toml:10]: Separate with spaces

**Comment by xiantang** [comment]
any update?

**Comment by kareemmahlees** [comment]
Sorry for the delay, I updated the tests hoping that they do a little more this time, unfortunately, I can't ensure that the tests are passing because there is a problem with `t.tempDir` on Windows specifically 😢

**Comment by kareemmahlees** [comment]
I already have the latest version ( 1.21 ) installed, Sorry for being not so helpful but I can start working on post cmd if that's okay!

**Comment by xiantang** [comment]
Keep going

**Comment by kareemmahlees** [comment]
sorry for the delay, been busy at work.
@xiantang could you check the latest changes?

**Comment by xiantang** [comment]
can let me check

File: air_example.toml [hunk 3]
```diff
@@ -6,8 +6,12 @@ root = "."
 tmp_dir = "tmp"
 
 [build]
+# Array of commands to run before each build
+pre_cmd = ["echo \"hello air\" > pre_cmd.txt"]
```
Comment (xiantang) [inline air_example.toml:10]: "echo 'hello air' > pre_cmd.txt"

Comment (kareemmahlees) [reply air_example.toml]: Done !

**Comment by xiantang** [comment]
LGTM merged!

**Comment by xiantang** [comment]
thank you!

**Comment by jackielii** [comment]
Hmm, I thought post_cmd would run each time after cmd is restarted... it's surprising to call it post_cmd, should've been `before_exit`

**Pull request merged**